### PR TITLE
fix(ci): add native macOS build paths for UDB dependencies (#2583)

### LINKED ISSUE #2583: udb gem installs Linux (ELF) z3 binary instead of macOS (Mach-O) on arm64
**Describe the bug**
udb installs Linux (ELF) z3 binary instead of macOS (Mach-O) on arm64

**To Reproduce**
Environment:

    macOS, Apple Silicon (arm64)
    Ruby 3.4.10
    udb gem 0.1.16, bundler 4.0.19

Steps to reproduce:

1. Fresh bundle install that pulls in the udb gem (0.1.16) on an arm64 Mac
2. Run any command that triggers UDB config validation (e.g. make)

**Text output/Error Messages**
Note: Absolute paths in the error message have been replaced with `~`.

```
$ make
Could not open library '~/.cache/udb/z3/z3-5.0.0/arm64/libz3.so': dlopen(~/.cache/udb/z3/z3-5.0.0/arm64/libz3.so, 0x0005): tried: '~/.cache/udb/z3/z3-5.0.0/arm64/libz3.so' (slice is not valid mach-o file), '/System/Volumes/Preboot/Cryptexes/OS/Users/user/.cache/udb/z3/z3-5.0.0/arm64/libz3.so' (no such file), '~/.cache/udb/z3/z3-5.0.0/arm64/libz3.so' (slice is not valid mach-o file).
```

**Additional context**
Workaround:
brew install z3
cp "$(brew --prefix z3)/lib/libz3.dylib" ~/.cache/udb/z3/z3-5.0.0/arm64/libz3.so

### PR DESCRIPTION
**Pull request by drona-gyawali**
Fixes #2583

## Problem
extconf.rb was downloading Linux ELF binaries (.so) on all platforms.
On macOS.  So, we are solving this by extending the build  pipeline

## Changes

**extconf.rb** : added host_os detection, all 4 binaries now use
OS-aware download URLs and correct library extension (.so/.dylib)

**update.sh** : OS detection added to all 4 build functions; on Linux
calls existing _with_docker.sh scripts; on Mac calls new native scripts

**4 new native macOS build scripts** : build_espresso_mac.sh,
build_must_mac.sh, build_eqntott_mac.sh, build_z3_mac.sh,  build
natively using the host toolchain (no Docker), produce real Mach-O
binaries, and verify with file() before finishing

**release_udb_deps.yml** : added macos-15 (x64) and macos-15-xlarge
(arm64) runners to all 4 build jobs

**z3_loader.rb** : fixed library_name being private while called
externally; fixed FFI override to use library_name() instead of
the raw incoming name

**Note:**
I didn't test this PR locally since it heavily involves Mac hardware, release files, etc. Just giving a heads-up!

### CODE REVIEW

File: .github/workflows/release_udb_deps.yml [hunk 1]
```diff
@@ -101,6 +101,10 @@ jobs:
             arch: x64
           - runner: ubuntu-24.04-arm
             arch: arm64
+          - runner: macos-15
+            arch: x64
+          - runner: macos-15-xlarge
+            arch: arm64
```
Comment (jordancarlin) [inline .github/workflows/release_udb_deps.yml]: Based on https://github.com/actions/runner-images#available-images
```suggestion
          - runner: macos-15-intel
            arch: x64
          - runner: macos-15
            arch: arm64
```

File: .github/workflows/release_udb_deps.yml [hunk 2]
```diff
@@ -116,10 +120,10 @@ jobs:
             ./bin/chore update eqntott -n
           fi
 
-  build-z3:
+  build-espresso:
```
Comment (jordancarlin) [inline .github/workflows/release_udb_deps.yml]: Please don't change the order of the tool builds. It makes this much harder to review because GitHub shows a bunch of changes that aren't significant.

File: tools/ruby-gems/udb/ext/udb_download/extconf.rb [hunk 3]
```diff
@@ -40,6 +40,20 @@ def find_executable(name)
           "Only x64 and arm64 are supported."
   end
 
+# ---- detect host OS ----
```
Comment (jordancarlin) [inline tools/ruby-gems/udb/ext/udb_download/extconf.rb]: The arch detection doesn't have this comment banner. Let's stay consistent.

File: tools/ruby-gems/udb/ext/udb_download/extconf.rb [hunk 4]
```diff
@@ -40,6 +40,20 @@ def find_executable(name)
...
+os_ext =
```
Comment (jordancarlin) [inline tools/ruby-gems/udb/ext/udb_download/extconf.rb]: Why `os_ext`? This isn't an extension, it is the os.

**Review by jordancarlin — CHANGES_REQUESTED** [review]
Are separate build scripts for macOS really needed? I assume most of the build process is the same for Linux and macOS. Can we just have a few ifs in the existing scripts instead?

**Comment by drona-gyawali** [comment]
> Are separate build scripts for macOS really needed? I assume most of the build process is the same for Linux and macOS. Can we just have a few ifs in the existing scripts instead?

Thanks for the feedback! I was also thinking along the same lines , the existing build script filenames say "with_docker" and they install a Linux-based toolchain (musl + AlmaLinux), which isn't supported on macOS runners. That was my reason for keeping them separate. I'm not sure we can cleanly stitch the two together, but happy to consolidate if you prefer.

File: .github/workflows/release_udb_deps.yml (hunk 1 above)
Comment (drona-gyawali) [reply .github/workflows/release_udb_deps.yml]: done , changes has been updated.

File: .github/workflows/release_udb_deps.yml (hunk 2 above)
Comment (drona-gyawali) [reply .github/workflows/release_udb_deps.yml]: sorry for that. now it looks clean, i guess.

File: tools/ruby-gems/udb/ext/udb_download/extconf.rb (hunk 3 above)
Comment (drona-gyawali) [reply tools/ruby-gems/udb/ext/udb_download/extconf.rb]: done

**Comment by jordancarlin** [comment]
> Thanks for the feedback! I was also thinking along the same lines , the existing build script filenames say "with_docker" and they install a Linux-based toolchain (musl + AlmaLinux), which isn't supported on macOS runners. That was my reason for keeping them separate. I'm not sure we can cleanly stitch the two together, but happy to consolidate if you prefer.

We should do a bit of an overhaul of those and dramatically simplify the builds, but this seems fine for now and I agree we should keep them separate.

File: bin/.chore/update.sh [hunk 5]
```diff
@@ -129,74 +148,89 @@ do_update_espresso() {
...
+    if [ "${current_os}" = "Mac" ]; then
+      "${UDB_ROOT}"/tools/scripts/build_espresso_mac.sh "${work_dir}/espresso-build" "${native_arch}" || exit 1
+    else
+      "${UDB_ROOT}"/tools/scripts/build_espresso_with_docker.sh "${work_dir}/espresso-build" "${native_arch}" || exit 1
+    fi
 
-    # Move the binary to the asset name expected by the gem
-    mv "${work_dir}/espresso-build/espresso" "${work_dir}/espresso-${native_arch}"
+    # include OS in asset name
+    mv "${work_dir}/espresso-build/espresso" "${work_dir}/espresso-${current_os}-${native_arch}"
```
Comment (jordancarlin) [inline bin/.chore/update.sh]: Revert this comment change.

File: bin/.chore/update.sh [hunk 6]
```diff
@@ -50,7 +50,7 @@ read_git_pinned_tool_version() {
     exit 1
   fi
 
-  commit="${version#${tool}-}"
+  commit="${version#"${tool}"-}"
```
Comment (jordancarlin) [inline bin/.chore/update.sh]: This quoting seems wrong.

File: bin/.chore/update.sh [hunk 7]
```diff
@@ -129,74 +148,89 @@ do_update_espresso() {
...
+    else
+      "${UDB_ROOT}"/tools/scripts/build_espresso_with_docker.sh "${work_dir}/espresso-build" "${native_arch}" || exit 1
+    fi
 
-    # Move the binary to the asset name expected by the gem
-    mv "${work_dir}/espresso-build/espresso" "${work_dir}/espresso-${native_arch}"
+    # include OS in asset name
+    mv "${work_dir}/espresso-build/espresso" "${work_dir}/espresso-${current_os}-${native_arch}"
 
-    # Generate checksum
```
Comment (jordancarlin) [inline bin/.chore/update.sh:138]: Why was this comment deleted?

File: tools/scripts/build_eqntott_mac.sh [hunk 8]
```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env bash
+# Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
```
Comment (jordancarlin) [inline tools/scripts/build_eqntott_mac.sh]: Include your own copyright in new files.

**Review by jordancarlin — CHANGES_REQUESTED** [review]
A few small changes. You also need to cryptographically sign all of your commits. Let me know if you need help with this.

Have you tested the macOS scripts on a Mac?

We should eventually also add some level of macOS CI to ensure everything continues to work, but that will need to come it a future PR because it will not work until these binaries are published, and that can't happen until this is merged.

File: bin/.chore/update.sh (hunk 6 above)
Comment (drona-gyawali) [reply bin/.chore/update.sh]: quoting inside parameter expansion like ${version#"${tool}"-} is valid syntax in Bash/POSIX, it basically disables pattern matching so variables with characters like * or ? are treated as literal strings. However, since ${tool} is just a plain name here, the quotes aren't strictly necessary. Reverting to ${version#${tool}-} makes sense to keep the code clean

File: bin/.chore/update.sh (hunk 7 above)
Comment (drona-gyawali) [reply bin/.chore/update.sh:138]: I removed it because I felt the echo "Generating checksum..." log line right below it made the comment redundant. However, I wil revert the change :)

**Comment by drona-gyawali** [comment]
> A few small changes. You also need to cryptographically sign all of your commits. Let me know if you need help with this.

Thanks, just recently knew about this i will add signed commits from onward.

> Have you tested the macOS scripts on a Mac?

To be frank, I don't have access to a Mac, so I wasn't able to test the scripts locally. I mentioned this in the  PR description. The scripts follow the same pattern as the existing Docker scripts and use standard Xcode/Clang toolchain commands that are available on macOS  runners. if you're able to test them on a Mac that would be very helpful :).

> We should eventually also add some level of macOS CI to ensure everything continues to work, but that will need to come it a future PR because it will not work until these binaries are published, and that can't happen until this is merged.

Understood,  agreed that macOS CI testing makes sense as a follow-up once the binaries are published. Happy to open a follow-up issue to track that if it would be helpful

**Comment by jordancarlin** [comment]
I don't want to merge this without testing it on a Mac first. I will try to do that soon, but it may have to wait a few days.

**Comment by drona-gyawali** [comment]
> I don't want to merge this without testing it on a Mac first. I will try to do that soon, but it may have to wait a few days.

Hi  jordan,

To test on actual macOS hardware, I set up a GitHub Actions test workflow running on macos-latest: https://github.com/drona-gyawali/mac-script-harness

While running the builds, I caught a few platform edge cases,  which I've updated in this PR.

I'm not 100% sure if a CI harness fully replaces a local manual run for you, but all four builds passed cleanly on the macOS runner, feel free to check the workflow logs!
